src/Grpc.StatusProto/README.md

```diff
@@ -0,0 +1,256 @@
+each object. The class `DetailsTypesRegistry` provides an `Unpack` method to decode the `Any` messages in the status details. C#'s switch pattern matching can be used to process the details.
+
+Example:
+```C#
+void PrintStatusDetails(RpcException ex)
+{
+    // Get the status from the RpcException
+    Google.Rpc.Status? rpcStatus = ex.GetRpcStatus(); // Extension method
+
+    if (rpcStatus != null)
+    {
+        // Decode each "Any" item from the details in turn
+        foreach (Any any in rpcStatus.Details)
+        {
+            // decode the message if it is one of those expected to be in the
+            // status details
+            IMessage msg = DetailsTypesRegistry.Unpack(any); 
+
+            switch (msg)
+            {
+                case null:
+                    // ignore
+                    Console.WriteLine("Unknown message type in Details");
+                    break;
+
+                case ErrorInfo errorInfo:
+                    Console.WriteLine($"ErrorInfo: Reason: {errorInfo.Reason}, Domain: {errorInfo.Domain}");
+                    foreach (var md in errorInfo.Metadata)
+                    {
+                        Console.WriteLine($"\tKey: {md.Key}, Value: {md.Value}");
+                    }
+                    break;
+
+                case BadRequest badRequest:
+                    Console.WriteLine("BadRequest:");
+                    foreach (BadRequest.Types.FieldViolation fv in badRequest.FieldViolations)
+                    {
+                        Console.WriteLine($"\tField: {fv.Field}, Description: {fv.Description}");
+                    }
+                    break;
+
+                // Other cases handled here ...
+            }
+        }
+    }
+
+```
+
+## Returning errors within gRPC streams
+
+The model described above allows you to return an error status when the gRPC call finishes. Sometimes
+when using gRPC streams that allow clients and servers to send multiple messages in a single call
+you may wish to return a status without terminating the call.
+
+To do this the definition of the stream of messages returned from the service should itself contain
+a `google.rpc.Status` message, for example:
+
+```protobuf
+service WidgetLookupProvider {
+    rpc streamingLookup(stream WidgetReq) returns (stream WidgetRsp) {}
+}
+
+message WidgetReq {
+    string widget_name = 1;
+}
+
+message WidgetRsp {
+    oneof message{
+        // details when ok
+        string widget_details = 1;
+        // or error details
+        google.rpc.Status status = 2;
+   }   
+}
+```
+
+Note: the  [status.proto](https://github.com/googleapis/googleapis/blob/master/google/rpc/status.proto)
+and [error_details.proto](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto)
+files are not currently provided in any NuGet packages. The text for those proto files can be copied from
+the given links and included in your project.
```

Comment:
nit: reword "The text for those proto files can be copied from
the given links and included in your project."

Reply:
This is interesting... do you think it would be beneficial to ship them in Google.Api.CommonProtos? I wouldn't be averse to that, but I'd need to consult the rest of the team.

Reply:
At this point, I'd say this wouldn't help much. I'd probably help if we had a general mechanism for shipping .proto file in nuget packages and Grpc.Tools could autogenerate source code for them (see https://github.com/grpc/grpc-dotnet/issues/183), but we currently don't have that.

Reply:
As it happens, we had a feature request for that last week: https://github.com/googleapis/gax-dotnet/issues/715
@jtattermusch If you believe it wouldn't actually help much, please could you comment on that issue?

Reply:
It is on my list to document a standard way/best practices for shipping proto files in NuGet packages so that Grpc.Tools can easily pick them up. Basically what @JamesNK does in the [Microsoft.AspNetCore.Grpc.JsonTranscoding](https://www.nuget.org/packages/Microsoft.AspNetCore.Grpc.JsonTranscoding) package.